server/src/main/java/io/littlehorse/server/auth/RequestAuthorizer.java

```diff
@@ -121,9 +120,22 @@ private void loadMetadataFromServiceClass(final Class<? extends BindableService>
             }
         }
 
-        private void verify(MethodDescriptor<?, ?> serviceMethod, Collection<ServerACLModel> acls) {
+        private void verify(MethodDescriptor<?, ?> serviceMethod, AuthorizationContext authContext) {
             String methodName = serviceMethod.getBareMethodName();
             AuthMetadata authMetadata = methodMetadata.get(methodName);
+
+            Collection<ServerACLModel> acls;
+
+            if (!authContext.perTenantAcls().isEmpty()) {
+                if (requiresClusterScopedResource(authMetadata)) {
+                    throw new PermissionDeniedException(
+                            "Insufficient ACLs. Access to Principals and Tenants can only be granted with Global ACLs");
+                }
+                acls = authContext.perTenantAcls();
```

Comment:
With this implementation, what is the reason to have different methods for `perTenantAcls` and global `acls`?

Reply:
Without both methods, `RequestAuthorizer` has no way of knowing if the ACLs present are from PerTenant or Global ACLs. So it cannot differentiate between `ACL_ALL_RESOURCES` in a global context or in a per-tenant context (where it blocks principal and tenant requests).

There are ways to refactor this perhaps, where the logic for deciding between PerTenant or Global is abstracted away, and maybe an Enum is passed on indicating the context.

Reply:
As far I understand, the Collection on [this line](https://github.com/littlehorse-enterprises/littlehorse/pull/1032/files/24eefaf32e4b84d2ad4743dba097dbd3f72b7cd4#diff-9937bc62b704d7fa071c63d6a7633f625fd209542ff341ed9aa045111fa4df80R127)  stores both global and perTenantAcls, so it seems that you need that only for this validation ([line #130](https://github.com/littlehorse-enterprises/littlehorse/pull/1032/files/24eefaf32e4b84d2ad4743dba097dbd3f72b7cd4#diff-9937bc62b704d7fa071c63d6a7633f625fd209542ff341ed9aa045111fa4df80R130)). is that correct?

Reply:
Yes, I believe. My logic is as follows:

1. If per-tenant ACLs exist on authContext AND the service being accessed requires cluster-scoped resources, reject the request
2. Otherwise, if per-tenant ACLs exist, use the per-tenant ACLs
3. If per-tenant ACLs don't exist in authContext, default to using global ACLs

Reply:
To satisfy this logic ^ we need two pieces of information:
1. Know whether or not the Principal has perTenantAcls for the tenant its making the request in
2. Know whether or not the service being accessed requires cluster-scoped resouces

Reply:
I think you can reject the request before the AuthContext is instantiated.
It sounds like we should not create an AuthContext If the requester has perTenantAcls AND the service being accessed requires cluster-scoped resources.

Reply:
My proposal is basically move step 1 to the point where we create AuthContext

Reply:
Ok, I think the ACLVerifier class is the place where we are checking the service class's @Authorize annotation metadata. And the annotation metadata is necessary to satisfy Required Information #2 "Know whether or not the service being accessed requires cluster-scoped resources".

So I propose we add a new public method to the ACLVerifier that can return information about the service being accessed, like whether or not it requires cluster scoped resources.

Reply:
Agree. I think you can add that method in the acl verifier that basically will help you to decide whether the request should be rejected or not before creating the AuthContext.